File: tests/model-library/credential-file.py
"""Separate binaries/processes; only synthetic credentials in a temporary directory."""
import pathlib, subprocess, tempfile, os, stat
root=pathlib.Path(__file__).resolve().parents[2]
with tempfile.TemporaryDirectory(prefix='u1-credentials-test-') as temp:
    base=pathlib.Path(temp); folder=base/'credentials'
    first=base/'version1'; second=base/'version2'
    for binary,flag in [(first,'-O0'),(second,'-O2')]:
        subprocess.run(['c++','-std=c++17',flag,str(root/'tests/model-library/credential-file.cpp'),'-o',str(binary)],check=True)
    def run(action,binary=first): subprocess.run([str(binary),action,str(folder)],check=True)
    run('missing');run('write');run('read',second);run('isolation',second);run('reject-invalid',second)
    assert stat.S_IMODE(folder.stat().st_mode)==0o700
    assert stat.S_IMODE((folder/'thingiverse.token').stat().st_mode)==0o600
    run('disconnect');run('disconnected',second)
    (folder/'thingiverse.token').write_text('bad token')
    run('reject-file');assert (folder/'thingiverse.token').read_text()=='bad token'
    (folder/'thingiverse.token').unlink()
    outside=base/'outside';outside.write_text('untouched')
    (folder/'thingiverse.token').symlink_to(outside)
    run('reject-file');assert outside.read_text()=='untouched'
    print('PASS permissions, corruption preservation, symlink rejection and version-independent storage')

File: tests/model-library/credential-file.cpp
#include "../../src/slic3r/GUI/LibraryCredentialFile.hpp"
#include <iostream>
int main(int argc, char** argv)
{
    using Slic3r::GUI::LibraryCredentialFile;
    if (argc != 3)
        return 2;
    LibraryCredentialFile store(argv[2]);
    std::string           action(argv[1]), value;
    const std::string     synthetic = "synthetic-test-only-not-a-key";
    bool                  ok        = false;
    try {
        if (action == "isolation") {
            LibraryCredentialFile other(argv[2], "makerworld.token");
            other.save(std::string(2048, 'x'));
            std::string second;
            ok = other.load(second) && second.size() == 2048 && store.load(value) && value == synthetic;
        }
        if (action == "write") {
            store.save(synthetic);
            ok = true;
        }
        if (action == "read")
            ok = store.load(value) && value == synthetic;
        if (action == "missing")
            ok = !store.load(value);
        if (action == "disconnect") {
            store.save("");
            ok = true;
        }
        if (action == "disconnected")
            ok = store.load(value) && value.empty();
        if (action == "reject-invalid") {
            try {
                store.save("invalid key");
            } catch (...) {
                ok = store.load(value) && value == synthetic;
            }
        }
        if (action == "reject-file") {
            try {
                store.load(value);
            } catch (...) {
                ok = true;
            }
        }
    } catch (...) {
        ok = false;
    }
    std::cout << (ok ? "PASS " : "FAIL ") << action << '\n';
    return ok ? 0 : 1;
}
